sdks/python/apache_beam/examples/complete/distribopt/distribopt/distribopt.py

```diff
@@ -0,0 +1,349 @@
+      # split of crops with highest number of options.
+      best_split = np.argsort([-len(rec['transport_costs']) for rec in records])[:2]
+      rec1 = records[best_split[0]]
+      rec2 = records[best_split[1]]
+
+      # Generate & emit all combinations
+      for a in rec1['transport_costs']:
+        if a[1]:
+          for b in rec2['transport_costs']:
+            if b[1]:
+              combination = [(rec1['crop'], a[0]), (rec2['crop'], b[0])]
+              yield pvalue.TaggedOutput('splitted', combination)
+
+      # Pass on remaining records
+      remaining = [rec for i, rec in enumerate(records) if i not in best_split]
+      yield pvalue.TaggedOutput('combine', remaining)
+
+  class GenerateMappings(beam.DoFn):
+    """
+    ParDo implementation to generate all possible assignments of crops to greenhouses.
+
+    Input: dict with crop, quantity and transport_costs keys, e.g.,
+    {
+        'crop': 'OP009',
+        'quantity': 102,
+        'transport_costs': [('A', None), ('B', 3), ('C', 8)]
+    }
+
+    Output: tuple (mapping_identifier, {crop -> greenhouse})
+    """
+
+    @staticmethod
+    def _coordinates_to_greenhouse(coordinates, greenhouses, crops):
+      # Map the grid coordinates back to greenhouse labels
+      arr = []
+      for coord in coordinates:
+        arr.append(greenhouses[coord])
+      return dict(zip(crops, np.array(arr)))
+
+    def process(self, element, records):
+      # Generate available greenhouses and grid coordinates for each crop.
+      grid_coordinates = []
+      for rec in records:
+        # Get indices for available greenhouses (w.r.t crops)
+        filtered = [i for i, available in enumerate(rec['transport_costs']) if available[1]]
+        grid_coordinates.append(filtered)
+
+      # Generate all mappings and .
+      grid = np.vstack(map(np.ravel, np.meshgrid(*grid_coordinates))).T
+      crops = [rec['crop'] for rec in records]
+      greenhouses = [rec[0] for rec in records[0]['transport_costs']]
+      mappings = []
+      for point in grid:
+        # translate back to greenhouse label
+        mapping = self._coordinates_to_greenhouse(point, greenhouses, crops)
+        assert all(rec[0] not in mapping for rec in element)
+        # include the incomplete mapping of 2 crops
+        mapping.update(element)
+        # include identifier
+        yield (uuid.uuid4().hex, mapping)
+
+  def expand(self, records):
+    o = (
+        records
+        | 'pair one' >> beam.Map(lambda x: (1, x))
+        | 'group all records' >> beam.GroupByKey()
+        | 'split one of' >> beam.ParDo(self.PreGenerateMappings())
+                              .with_outputs('splitted', 'combine')
+    )
+
+    # Create mappings, and prevent fusion (this limits the parallelization
+    # in the optimization step)
+    mappings = (
+        o.splitted
+        | 'create mappings' >> beam.ParDo(self.GenerateMappings(), pvalue.AsSingleton(o.combine))
+        | 'prevent fusion' >> beam.Reshuffle()
```

Comment:
Why are we preventing shuffle here?

It is hard to see why it is needed when pipeline is composed in multiple places. Consider moving pipeline construction into a single place.